Undefined behavior warning: `Triangulation::execute_refinement_isotropic/execute_coarsening`

<!--
Thanks for reporting an issue with deal.II! This HTML comment won't be visible
in the issue tracker but is present here to help you write a better description
of the problem you encountered.

If you have questions about deal.II, please post them on the mailing list:

   https://groups.google.com/g/dealii

Otherwise:
- Do you encounter a compilation failure? Please attach the detailed.log file
  present in the build directory here. Any additional compiler messages will also
  be helpful!

- Does something crash (e.g., an example) that shouldn't? Please include any relevant
  terminal output as well as the detailed.log file generated in your build
  directory (should you have one).

- Is there a missing feature you want to implement? Please briefly describe it and
  the problem it solves.

- Did you find some other issue with deal.II? Summarize the problem here.
-->

With GCC 12.1.1 I get the following warning

```
/home/dealii/source/grid/tria.cc: In function ‘static typename dealii::Triangulation<3, spacedim>::DistortedCellList dealii::internal::TriangulationImplementation::Implementation::execute_refinement_isotropic(dealii::Triangulation<3, spacedim>&, bool) [with int spacedim = 3]’:
/home/dealii/source/grid/tria.cc:5697:43: warning: iteration 12 invokes undefined behavior [-Waggressive-loop-optimizations]
 5697 |                         raw_line_iterator line(&triangulation,
      |                                           ^~~~
/home/dealii/source/grid/tria.cc:5695:48: note: within this loop
 5695 |                     for (unsigned int l = 0; l < hex->n_lines(); ++l)
```
and 

```
/home/dealii/source/grid/tria.cc: In member function ‘void dealii::Triangulation<<anonymous>, <anonymous> >::execute_coarsening() [with int dim = 3; int spacedim = 3]’:
/home/dealii/source/grid/tria.cc:14809:32: warning: iteration 12 invokes undefined behavior [-Waggressive-loop-optimizations]
14809 |               ++line_cell_count[line_indices[l]];
      |                 ~~~~~~~~~~~~~~~^
/home/dealii/source/grid/tria.cc:14808:40: note: within this loop
14808 |             for (unsigned int l = 0; l < cell->n_lines(); ++l)
```
Could it be that an array access might be out of bounds in iteration 12? I quickly went through the code but unfortunately could not identify the issue. It might be related to https://github.com/dealii/dealii/pull/14221, so maybe @kronbichler you know immediately if it is relevant and what it might cause.

--- a/include/deal.II/grid/reference_cell.h
+++ b/include/deal.II/grid/reference_cell.h
@@ -1007,7 +1007,7 @@ ReferenceCell::n_lines() const
     return 12;
 
   Assert(false, ExcNotImplemented());
-  return numbers::invalid_unsigned_int;
+  return 0;
 }
 
 

